iroh-net/src/net/netmon/actor.rs

```diff
@@ -0,0 +1,257 @@
+                    trace!("network activity detected");
+                    last_event.replace(false);
+                    debounce_interval.reset_immediately();
+                }
+                Some(msg) = self.actor_receiver.recv() => match msg {
+                    ActorMessage::Subscribe(callback, s) => {
+                        let token = self.next_callback_token();
+                        self.callbacks.insert(token, Arc::new(callback));
+                        s.send(token).ok();
+                    }
+                    ActorMessage::Unsubscribe(token, s) => {
+                        self.callbacks.remove(&token);
+                        s.send(()).ok();
+                    }
+                },
+                else => {
+                    break;
+                }
+            }
+        }
+    }
+
+    fn next_callback_token(&mut self) -> CallbackToken {
+        let token = CallbackToken(self.callback_token);
+        self.callback_token += 1;
+        token
+    }
+
+    async fn handle_potential_change(&mut self, time_jumped: bool) -> Result<()> {
+        info!("potential change");
+
+        let new_state = State::new().await;
+        let old_state = &self.interface_state;
+
+        // No major changes, continue on
+        if !time_jumped && old_state == &new_state {
+            debug!("no changes detected");
+            return Ok(());
+        }
+
+        let mut is_major = is_major_change(old_state, &new_state);
+        // Check for time jumps
+        if !is_major && time_jumped {
+            is_major = true;
+        }
+
+        if is_major {
+            self.interface_state = new_state;
+        }
+
+        debug!("triggering {} callbacks", self.callbacks.len());
+        for cb in self.callbacks.values() {
+            let cb = cb.clone();
+            tokio::task::spawn(async move {
+                cb(is_major).await;
+            });
+        }
+
+        Ok(())
+    }
+
+    /// Reports whether wall time jumped more than 150%
+    /// of `POLL_WALL_TIME_INTERVAL`, indicating we probably just came out of sleep.
+    fn check_wall_time_advance(&mut self) -> bool {
+        let now = Instant::now();
+        let jumped = if let Some(elapsed) = now.checked_duration_since(self.wall_time) {
+            elapsed > POLL_WALL_TIME_INTERVAL * 3 / 2
+        } else {
+            false
+        };
+
+        self.wall_time = now;
+        jumped
+    }
+}
+
+fn is_major_change(s1: &State, s2: &State) -> bool {
+    if s1.have_v6 != s2.have_v6
+        || s1.have_v4 != s2.have_v4
+        || s1.is_expensive != s2.is_expensive
+        || s1.default_route_interface != s2.default_route_interface
+        || s1.http_proxy != s2.http_proxy
+        || s1.pac != s2.pac
+    {
+        return true;
+    }
+
+    for (iname, i) in &s1.interface {
+        let Some(ips) = s1.interface_ips.get(iname) else {
+            // inconsistent dataset, ignore
+            continue;
+        };
```

Comment:
it's out of context for the current pr but it looks like we could store the ips along with the interfaces, saving us storing the name twice every time?

Reply:
removed, much cleaner now